libraries/chain/db_update.cpp

```diff
@@ -46,21 +47,23 @@ void database::update_global_dynamic_data( const signed_block& b )
       dynamic_global_property_id_type(0)(*this);
 
    uint32_t missed_blocks = get_slot_at_time( b.timestamp );
-   assert( missed_blocks != 0 );
+   FC_ASSERT( missed_blocks != 0, "Trying to push double-produced block onto current block?!" );
    missed_blocks--;
-   for( uint32_t i = 0; i < missed_blocks; ++i ) {
-      const auto& witness_missed = get_scheduled_witness( i+1 )(*this);
-      if(  witness_missed.id != b.witness ) {
-         /*
-         const auto& witness_account = witness_missed.witness_account(*this);
-         if( (fc::time_point::now() - b.timestamp) < fc::seconds(30) )
-            wlog( "Witness ${name} missed block ${n} around ${t}", ("name",witness_account.name)("n",b.block_num())("t",b.timestamp) );
-            */
-
-         modify( witness_missed, [&]( witness_object& w ) {
-           w.total_missed++;
-         });
-      } 
+   const auto& witnesses = witness_schedule_id_type()(*this).current_shuffled_witnesses;
+   uint64_t current_slot = get_dynamic_global_properties().current_aslot % witnesses.size();
+   const uint32_t missed_rounds = missed_blocks / witnesses.size();
+   if( missed_rounds > 0 )
+       for( const auto& witness : witnesses )
```

Comment:
I'd like to skip updating `total_missed` data for `b.witness` here, aka keep current behavior.

Rationale: the witness that produced the first block after revived a halted chain will get a good-looking block missing number which is somewhat a prize. IMHO that's why we didn't "fix" it before.

Reply:
I'd prefer to fix this, at least for the future.
* The first witness to produce a block after a large gap has still missed blocks just like everyone else. The missed blocks count is wrong and therefore misleading if we don't count these.
* In case of a chain halt, we need *cooperation* among witnesses, not *competition*. "Rewarding" one witness incentivizes *competition* and is unfair on all the other *cooperating* witnesses.

Reply:
Possible alternative: if missed_blocks >= witnesses.size don't count misses for anyone.

Reply:
The alternative makes some sense. A chain halt is not any witness's fault.

Note: need to make sure that after a maintenance block, the size of active witness list can change and the list will be re-shuffled.

Reply:
Re-shuffling happens after counting missed blocks, and does not happen while blocks are being missed, so using current list for the whole range is correct.